### ISSUE TITLE
Identical return type considered different in redeclaration with trailing return type and enough templates

### ISSUE THREAD

**Issue by davidstone** [llvm/llvm-project#91650]
The following valid code:

```c++
template<bool>
struct enable_if {
	using type = void;
};

template<typename>
struct trait {
	static constexpr bool value = false;
};

template<typename T>
struct s {
  typename enable_if<trait<T>::value>::type assign();
};

template<typename T>
auto s<T>::assign() -> typename enable_if<trait<T>::value>::type {
}
```

is rejected by clang with

```console
<source>:17:12: error: return type of out-of-line definition of 's::assign' differs from that in the declaration
   17 | auto s<T>::assign() -> typename enable_if<trait<T>::value>::type {
      |            ^
<source>:13:45: note: previous declaration is here
   13 |   typename enable_if<trait<T>::value>::type assign();
      |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
1 error generated.
Compiler returned: 1
```

This occurred on clang trunk very recently, so it's not yet showing up on Godbolt: https://godbolt.org/z/KTE7dvcjd

@sdkrystian: It was introduced by 62b5b61f436add042d8729dc9837d055613180d9.

**Comment by davidstone** [comment]
The net effect of this is that `#include <boost/static_string.hpp>` fails to compile (which is in turn used by `boost::beast`.

**Comment by sdkrystian** [comment]
@davidstone I'll be merging #91620 shortly, which resolves the issue

**Comment by sdkrystian** [comment]
Closed via #91620
